sources/packages/backend/apps/workers/src/services/disbursement-schedule/disbursement-schedule.service.ts

```diff
@@ -0,0 +1,175 @@
+import { Injectable } from "@nestjs/common";
+import {
+  ApplicationStatus,
+  DisbursementSchedule,
+  MSFAANumber,
+  OfferingIntensity,
+  RecordDataModelService,
+} from "@sims/sims-db";
+import { CustomNamedError } from "@sims/utilities";
+import { DataSource, IsNull, Not } from "typeorm";
+import { MSFAANumberService } from "..";
+import {
+  DISBURSEMENT_MSFAA_ALREADY_ASSOCIATED,
+  DISBURSEMENT_NOT_FOUND,
+} from "../../constants";
+
+@Injectable()
+export class DisbursementScheduleService extends RecordDataModelService<DisbursementSchedule> {
+  constructor(
+    dataSource: DataSource,
+    private readonly msfaaNumberService: MSFAANumberService,
+  ) {
+    super(dataSource.getRepository(DisbursementSchedule));
+  }
+
+  /**
+   * Associates an MSFAA number to the disbursement(s) checking
+   * whatever is needed to create a new MSFAA or use an
+   * existing one instead.
+   * @param assessmentId assessment id of the workflow instance.
+   */
+  async associateMSFAANumber(assessmentId: number): Promise<void> {
+    const disbursements = await this.repo.find({
+      select: {
+        id: true,
+        studentAssessment: {
+          id: true,
+          application: { id: true, student: { id: true } },
+          offering: { id: true, offeringIntensity: true, studyStartDate: true },
+        },
+        msfaaNumber: {
+          id: true,
+        },
+      },
+      relations: {
+        studentAssessment: { application: { student: true }, offering: true },
+        msfaaNumber: true,
+      },
+      where: {
+        studentAssessment: { id: assessmentId },
+      },
+    });
+
+    const [firstDisbursement] = disbursements;
+
+    if (!firstDisbursement) {
+      throw new CustomNamedError(
+        "Disbursement not found.",
+        DISBURSEMENT_NOT_FOUND,
```

Comment:
Just for my understanding: if the disbursement is not found here, it will try to associate again in the next scheduled time, right? How many times the queue tries to associate it?

Reply:
Associate disbursement here is the state of workflow which happens after disbursements are saved for an assessment.

![image](https://user-images.githubusercontent.com/54600590/229014825-d21ebc93-9a24-4dbd-a1aa-972d85d42ee1.png)

So the expectation at this state is there must be disbursement(s) available for the assessment to make the association happen. 

There must never be a situation when trying to associate msfaa, the disbursements are not found unless there is hidden bug in workflow(for e.g. Save disbursement having an error but not failing and moving to next stage).